core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -864,15 +925,37 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
     }
 
     /**
+     * Deletes a snapshot from the repository, looking up the {@link Snapshot} reference before deleting.
+     * If the snapshot is still running cancels the snapshot first and then deletes it from the repository.
```

Comment:
I can do both, because while it is the behavior that *if* there are two snapshots of the same name, one of them must be in the cluster state as "in progress" that is in the process of being removed from the cluster state, we can add the removal here as a safeguard because in `remoteSnapshotFromClusterState`, if the snapshot is not found in the cluster state (i.e. the `deleteSnapshot` removed it), then its essentially a no-op.